cpp/src/merge/merge.cu

```diff
@@ -342,17 +339,65 @@ std::unique_ptr<column> column_merger::operator()<cudf::dictionary32>(
 {
   auto result = cudf::dictionary::detail::merge(
     cudf::dictionary_column_view(lcol), cudf::dictionary_column_view(rcol), row_order_, stream, mr);
+
   // set the validity mask
   if (lcol.has_nulls() || rcol.has_nulls()) {
     auto merged_view = result->mutable_view();
-    materialize_bitmask(lcol, rcol, merged_view, row_order_.data(), stream);
+    materialize_bitmask(
+      lcol, rcol, merged_view.null_mask(), merged_view.size(), row_order_.data(), stream);
   }
   return result;
 }
 
+// specialization for structs
+template <>
+std::unique_ptr<column> column_merger::operator()<cudf::struct_view>(
+  column_view const& lcol,
+  column_view const& rcol,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) const
+{
+  // merge each child.
+  std::vector<std::unique_ptr<column>> merged_children;
+  merged_children.reserve(lcol.num_children());
+
+  structs_column_view lhs(lcol);
+  structs_column_view rhs(rcol);
+
+  column_merger merger{row_order_};
+  auto iter = thrust::make_counting_iterator(0);
+  std::transform(
+    iter, iter + lhs.num_children(), std::back_inserter(merged_children), [&](size_type i) {
+      return cudf::type_dispatcher<dispatch_storage_type>(
+        lhs.child(i).type(), merger, lhs.get_sliced_child(i), rhs.get_sliced_child(i), stream, mr);
+    });
+
+  auto const merged_size = lcol.size() + rcol.size();
+
+  // materialize the output buffer
+  rmm::device_buffer validity =
+    lcol.has_nulls() || rcol.has_nulls()
+      ? create_null_mask(merged_size, mask_state::UNINITIALIZED, stream, mr)
+      : rmm::device_buffer{};
+  if (lcol.has_nulls() || rcol.has_nulls()) {
+    materialize_bitmask(lcol,
+                        rcol,
+                        static_cast<bitmask_type*>(validity.data()),
+                        merged_size,
+                        row_order_.data(),
+                        stream);
+  }
```

Comment:
I think we should use `is_nullable` here and let the caller be responsible for introspecting the data. If either of the inputs are nullable, the output should be nullable, regardless of whether there any invalid elements.

Reply:
That's definitely not consistent with how cudf works though.  We try and drop validity whenever possible.

Reply:
I think it's a mixed bag whether we do or not, but I think there's a strong argument for letting the user decide. Everywhere else we strongly prefer to avoid data introspection, and null-counting is data introspection. Switching to `is_nullable` means we are no longer introspecting the validity.

Reply:
@jrhemstad   Pinging Jake.

Reply:
Approving because there's no point in blocking on this.

Reply:
just want to start a discussion.

Reply:
I think the meta here is that we should be expecting null count to be already computed more often than not, so at least on the performance side it theoretically free.